### ISSUE TITLE
New Rule: prefer using `.at()`

### ISSUE THREAD

**Issue by aladdin-add** [eslint/eslint#15097]
### Rule details

prefer using `.at()` to all the basic indexable classes (Array, String, TypedArray)

### Related ECMAScript feature

https://github.com/tc39/proposal-relative-indexing-method

### What type of rule is this?

Suggests an alternate way of doing something

### Example code

incorrect examples:
```js
var arr = [....];
var last = arr[arr.length - 1];
```

correct examples:
```js
var arr = [....];
var last = arr.at(-1);
```

### Why should this rule be in the core instead of a plugin?

it's not platform-specific.

### Participation

- [ ] I am willing to submit a pull request to implement this rule.

### Additional comments

it has been stage-4.

**Comment by mdjermanovic** [comment]
That would be great for arrays, but how can we avoid false positives on array-like objects?

**Comment by aladdin-add** [comment]
yes, like [`array-callback-return`](https://eslint.org/docs/rules/array-callback-return), it will have a few false positives.

**Comment by bradzacher** [comment]
Yes, this would be easy to do with a rule using type information.
Grab computed member expressions and then check the type of the `.object` node.

NGL though you'd get 99.99% of the way there with a non-type-aware rule -- it'd only false-positive on objects that both have a `.length` prop AND have numbered indicies which would be super rare.

**Comment by ljharb** [comment]
Since String, Array, and Typed Arrays qualify but all have `.at`, that would basically narrow down the remainder to NodeLists (altho hopefully the web will add `.at` to them), and `arguments` objects (which probably aren't all that rare?)

**Comment by mdjermanovic** [comment]
We can detect directly accessed `arguments`, but `NodeList` would be a problem.

The `.at()` proposal lists [several interfaces](https://github.com/tc39/proposal-relative-indexing-method#convertable-interfaces) that also represent array-like objects, so those objects don't seem to be that rare.

I think this rule without types cannot be safely used in all environments, and therefore doesn't qualify for a core rule.

**Comment by bmish** [comment]
Unicorn rule for reference: [unicorn/prefer-at](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/prefer-at.md)

**Comment by nzakas** [comment]
> I think this rule without types cannot be safely used in all environments, and therefore doesn't qualify for a core rule.

I agree. In general, instance method related rules are problematic because type information is required to be safe. 

Given that there’s already a plugin rule filling this gap, I think we should decline.
